### PR TITLE
Bootstrap device enrollment instructions UI

### PR DESCRIPTION
**Pull request by jhe-okta**
## Description:

While authenticator is required to access the resource but the device does not have OV or OV is not signed in. We want to show these new UIs to instruct enduser on how to proceed base on the API design:

UI: https://oktawiki.atlassian.net/wiki/spaces/eng/pages/1301556200/Design+ODA+MDM+remediation+flow#Remediation-Flow
API: 
https://oktawiki.atlassian.net/wiki/spaces/eng/pages/1301556200/Design+ODA+MDM+remediation+flow#MDM-Remediation-API
https://oktawiki.atlassian.net/wiki/spaces/eng/pages/1301556200/Design+ODA+MDM+remediation+flow#ODA-Remediation-API

The UI is a deadend of the sign in flow, but not exactly like a terminal state. After discussed with IDX team, we decided not to use `messages` object in the IDX API response but to introduce a new object `deviceEnrollment`. 

This PR only focuses on bootstrapping the SIW flow with IDX API, the UI itself needs to be further polished. Please ignore the UI inconsistency with the mock for now.

## PR Checklist

- [ ] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Issue:

- [OKTA-336701](https://oktainc.atlassian.net/browse/OKTA-336701)

### CODE REVIEW

File: src/v2/ion/RemediationConstants.js [hunk 1]
```diff
@@ -41,6 +41,7 @@ const FORMS = {
   DEVICE_APPLE_SSO_EXTENSION: 'device-apple-sso-extension',
   CANCEL_TRANSACTION: 'cancel-transaction',
   LAUNCH_AUTHENTICATOR: 'launch-authenticator',
+  DEVICE_ENROLLMENT: 'device-enrollment',
```
Comment (pradeepdewda-okta) [inline src/v2/ion/RemediationConstants.js:44]: let's change this to be more specific.. `DEVICE_ENROLLMENT_TERMINAL_VIEW`

File: src/v2/ion/responseTransformer.js [hunk 2]
```diff
@@ -77,6 +77,11 @@ const getRemediationValues = (idx) => {
         // messages generically.
         value: [],
       });
+    } else if (idx.context.deviceEnrollment) {
```
Comment (pradeepdewda-okta) [inline src/v2/ion/responseTransformer.js:80]: I am fine this approach, it's reasonable considering end state for device enrollment but still a special type.

Just add a comment here..
// no remediation for device enrollment terminal state

File: src/v2/ion/responseTransformer.js [hunk 3]
```diff
@@ -77,6 +77,11 @@ const getRemediationValues = (idx) => {
...
+      remediationValues.push({
+        name: RemediationForms.DEVICE_ENROLLMENT,
+        value: [],
+      });
```
Comment (pradeepdewda-okta) [inline src/v2/ion/responseTransformer.js:84]: Also add `deviceEnrollment` to the `example` we have at line 164 in this file.

```
 Example: {
 idx: {},
 messages: {}
...
...
 deviceEnrollment: {}
  }	 
```

File: src/v2/view-builder/ViewFactory.js [hunk 4]
```diff
@@ -158,6 +159,11 @@ const VIEWS_MAPPING = {
   [RemediationForms.REDIRECT_IDP]: {
     [DEFAULT]: IdentifierView,
   },
+
+  [RemediationForms.DEVICE_ENROLLMENT]: {
+    [DEFAULT]: DeviceEnrollmentView,
```
Comment (pradeepdewda-okta) [inline src/v2/view-builder/ViewFactory.js:164]: `DeviceEnrollmentTerminalView`

File: src/v2/view-builder/views/DeviceEnrollmentView.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      'Download Okta Verify' : 'Additional setup required';
+  },
+
+  initialize () {
+    BaseForm.prototype.initialize.apply(this, arguments);
+    const deviceEnrollment = this.options.appState.get('deviceEnrollment');
+    this.enrollmentType = (deviceEnrollment.name || '').toLowerCase(); // oda/mdm
+    switch (this.enrollmentType) {
+    case ODA:
+      this.add(View.extend({
```
Comment (pradeepdewda-okta) [inline src/v2/view-builder/views/DeviceEnrollmentView.js:25]: I would extract these views from here and assign them as constant before we extend BaseForm

**Review by pradeepdewda-okta — COMMENTED** [review]
Looks fine. I feel this is good approach.

File: src/v2/ion/RemediationConstants.js (hunk 1 above)
Comment (haishengwu-okta) [reply src/v2/ion/RemediationConstants.js:44]: this meant for form name hence shall be same to form name unless we change form name in api.

File: src/v2/view-builder/views/DeviceEnrollmentView.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+const MDM = 'mdm';
+const IOS = 'ios';
+const ANDROID = 'android';
+
+const Body = BaseForm.extend({
+  noButtonBar: true,
+
+  title () {
+    return this.enrollmentType === ODA ?
+      'Download Okta Verify' : 'Additional setup required';
```
Comment (haishengwu-okta) [inline src/v2/view-builder/views/DeviceEnrollmentView.js:16]: several hard-code strings in this file

File: src/v2/view-builder/views/DeviceEnrollmentView.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  },
+
+  initialize () {
+    BaseForm.prototype.initialize.apply(this, arguments);
+    const deviceEnrollment = this.options.appState.get('deviceEnrollment');
+    this.enrollmentType = (deviceEnrollment.name || '').toLowerCase(); // oda/mdm
```
Comment (haishengwu-okta) [inline src/v2/view-builder/views/DeviceEnrollmentView.js:22]: did you mean `toUpperCase` ?

File: src/v2/view-builder/views/DeviceEnrollmentView.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        },
+      }));
+      break;
+    }
+  },
+});
+
+export default BaseView.extend({
+  Body,
+  Footer: '',
```
Comment (haishengwu-okta) [inline src/v2/view-builder/views/DeviceEnrollmentView.js:79]: why this change `Footer: ''` ?

File: src/v2/ion/RemediationConstants.js (hunk 1 above)
Comment (haishengwu-okta) [reply src/v2/ion/RemediationConstants.js:44]: Miss that this is the Form we created only for UI. Make sense to change to `DEVICE_ENROLLMENT_TERMINAL_VIEW`

File: src/v2/view-builder/views/DeviceEnrollmentTerminalView.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        },
+      }));
+      break;
+    }
+  },
+});
+
+export default BaseView.extend({
+  Body,
+  Footer: '', // Sign out link appears in the footer if a cancel object exists in API response
```
Comment (haishengwu-okta) [inline src/v2/view-builder/views/DeviceEnrollmentTerminalView.js:79]: Is it possible to tweak API to avoid cancel object?

File: src/v2/view-builder/views/DeviceEnrollmentTerminalView.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
+import { createCallout, View } from 'okta';
+import hbs from 'handlebars-inline-precompile';
+import BaseView from '../internals/BaseView';
+import BaseForm from '../internals/BaseForm';
+
+const ODA = 'oda';
```
Comment (haishengwu-okta) [inline src/v2/view-builder/views/DeviceEnrollmentTerminalView.js:6]: Be more specific like 
- `ENROLLMENT_TYPE_ODA`, etc
- `PLATFORM_IOS`, etc

File: playground/mocks/data/idp/idx/mdm-enrollment.json [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+      "vendor": "Airwatch",
+      "enrollmentLink": "https://sampleEnrollmentlink.com"
+    }
+  },
+  "cancel": {
+      "rel": [
+          "create-form"
+      ],
+      "name": "cancel",
+      "href": "https://idx.okta1.com/idp/idx/cancel",
```
Comment (haishengwu-okta) [inline playground/mocks/data/idp/idx/mdm-enrollment.json:19]: Change link  `https://idx.okta1.com` to `http://localhost:3000` just be consistent.

**Comment by haishengwu-okta** [comment]
moved to https://github.com/okta/okta-signin-widget/pull/1470
